baikal/_core/pprint.py

```diff
@@ -0,0 +1,526 @@
+# THEREOF, EVEN IF ADVISED OF THE POSSIBILITY THEREOF.
+
+# 6. This License Agreement will automatically terminate upon a material
+# breach of its terms and conditions.
+
+# 7. Nothing in this License Agreement shall be deemed to create any
+# relationship of agency, partnership, or joint venture between PSF and
+# Licensee. This License Agreement does not grant permission to use PSF
+# trademarks or trade name in a trademark sense to endorse or promote products
+# or services of Licensee, or any third party.
+
+# 8. By copying, installing or otherwise using Python, Licensee agrees to be
+# bound by the terms and conditions of this License Agreement.
+
+
+# Brief summary of changes to original code:
+# - "compact" parameter is supported for dicts, not just lists or tuples
+# - estimators have a custom handler, they're not just treated as objects
+# - long sequences (lists, tuples, dict items) with more than N elements are
+#   shortened using ellipsis (', ...') at the end.
+
+# Brief summary of changes to the scikit-learn code:
+# - Prints changed parameters only, by default, unless sklearn is importable and
+#   print_changed_only is set to False in the global config.
+# - It has a local definition of is_scalar_nan.
+# - Includes the step-specific parameters in the repr.
+# - Revise logic to extract default parameters from the step signature.
+# - Migrate brute-force ellipsis post-processing logic from BaseEstimator.__repr__
+
+from inspect import signature
+import numbers
+import pprint
+import re
+from collections import OrderedDict
+
+import numpy as np
+
+from baikal import Step, get_config
+
+
+def is_scalar_nan(x):
+    """Tests if x is NaN
+
+    This function is meant to overcome the issue that np.isnan does not allow
+    non-numerical types as input, and that np.nan is not np.float('nan').
+
+    Parameters
+    ----------
+    x : any type
+
+    Returns
+    -------
+    boolean
+
+    Examples
+    --------
+    >>> is_scalar_nan(np.nan)
+    True
+    >>> is_scalar_nan(float("nan"))
+    True
+    >>> is_scalar_nan(None)
+    False
+    >>> is_scalar_nan("")
+    False
+    >>> is_scalar_nan([np.nan])
+    False
+    """
+    # convert from numpy.bool_ to python bool to ensure that testing
+    # is_scalar_nan(x) is True does not fail.
+    return bool(isinstance(x, numbers.Real) and np.isnan(x))
+
+
+class KeyValTuple(tuple):
+    """Dummy class for correctly rendering key-value tuples from dicts."""
+    def __repr__(self):
+        # needed for _dispatch[tuple.__repr__] not to be overridden
+        return super().__repr__()
+
+
+class KeyValTupleParam(KeyValTuple):
+    """Dummy class for correctly rendering key-value tuples from parameters."""
+    pass
+
+
+def _get_params(step, changed_only):
+    params = step.get_params(deep=False) if hasattr(step, "get_params") else {}
+    if changed_only and params:
+        init_func = step._get_super_class_with_init().__init__
```

Comment:
The pretty-printer from sklearn does `estimator.__init__` directly, causing step classes to fail. So a custom implementation was needed.